# 5차시 강사용 Notebook — 데이터 훑어보기와 정리(EDA)

**시연 전 안내**: Orange3로 File → Data Table → Feature Statistics를 먼저 보여준 뒤 이 노트북으로 넘어옵니다.

## 1단계. 데이터 읽고 결측값 확인하기

**진행 방법**: 강의 ② 절 순서대로 "아주 작은 표"(toy) → 미리 맛보기 네 줄 → `describe()` → `isna().sum()`을 한 셀씩 실행한다. 미리 맛보기는 결과만 눈으로 보게 하고, 의미는 2~4단계에서 반도체 데이터로 다시 짚는다.

**설명 포인트**: `describe()` 결과에서 `count`가 113보다 작은 열(결측 있음), `mean`(304.94)이 `50%`(300.45)보다 눈에 띄게 큰 온도_섭씨(극단적으로 큰 값 의심), 음수인 압력_Pa `min`을 학생이 스스로 짚게 한다. `합격여부`는 1/-1 코드라 평균(0.59)을 그대로 읽으면 뜻이 없다는 점도 함께 말한다.

### 따라 해보기 — 아주 작은 표로 오늘 배울 것 미리 보기
오늘 배울 "결측값·이상값 찾기"를 113행짜리 반도체 데이터로 연습하기 전에, 학생 5명 점수로 먼저 감을 잡습니다
(강의 자료 "먼저 아주 작은 표로 오늘 배울 것 미리 보기"). 한 명은 점수가 비어 있고(결측값),
한 명은 550점으로 말이 안 되는 값(이상값)입니다.

In [1]:
# 딕셔너리로 학생 5명짜리 작은 표를 직접 만들어 toy에 저장하고 출력합니다.
# 실행하면 이름·점수 2개 열이 보입니다 — 하은의 점수는 NaN(비어 있음), 예린의 점수는 550.0(말이 안 되는 값)입니다.
# 점수 열에 None이 섞이면 나머지 숫자도 88.0처럼 소수점이 붙어 나옵니다(NaN이 소수형 값이기 때문입니다).
import pandas as pd

toy = pd.DataFrame({
    "이름": ["지민", "하은", "서준", "예린", "도윤"],
    "점수": [88, None, 92, 550, 79],   # None을 넣으면 결측값(NaN)이 된다
})
print(toy)

   이름     점수
0  지민   88.0
1  하은    NaN
2  서준   92.0
3  예린  550.0
4  도윤   79.0


아래 네 셀은 강의 자료의 "오늘 배울 것 미리 맛보기" 네 줄입니다. 하나씩 실행해 결과만 눈으로 훑어보세요 —
정확한 의미는 뒤 단계에서 반도체 데이터로 하나씩 자세히 배웁니다.

In [2]:
# 미리 맛보기 1 — 결측값 개수 세기: 열마다 비어 있는 칸이 몇 개인지 셉니다.
# 실행하면 이름 0, 점수 1이 나옵니다 — 하은의 NaN 1건이 잡힌 것입니다.
# isna(): 칸마다 비어 있으면 True, 값이 있으면 False로 바꿉니다. 이어지는 .sum()이 True를 1로 세어 열별로 더합니다.
toy.isna().sum()

이름    0
점수    1
dtype: int64

In [3]:
# 미리 맛보기 2 — 결측값이 있는 행 제거: 비어 있는 칸이 하나라도 있는 행을 뺀 결과를 보여줍니다.
# 실행하면 하은(1번 행)이 사라진 4행짜리 표가 보입니다. 결과를 저장하지 않았으므로 원본 toy는 그대로입니다.
# dropna(): 결측값이 있는 행을 지운 "새 결과"를 돌려줍니다.
toy.dropna()

,이름,점수
0,지민,88.0
2,서준,92.0
3,예린,550.0
4,도윤,79.0


In [4]:
# 미리 맛보기 3 — 결측값을 중앙값으로 채우기: 점수 열의 빈칸을 점수들의 중앙값으로 채운 결과를 보여줍니다.
# 실행하면 1번(하은) 자리에 90.0이 들어간 점수 5개가 나옵니다(79, 88, 92, 550의 중앙값 = 88과 92의 가운데 = 90).
# median(): 값을 크기순으로 늘어놓았을 때 가운데 값, fillna(값): 빈칸을 그 값으로 채운 새 결과를 돌려줍니다.
toy["점수"].fillna(toy["점수"].median())

0     88.0
1     90.0
2     92.0
3    550.0
4     79.0
Name: 점수, dtype: float64

In [5]:
# 미리 맛보기 4 — 조건 검색으로 이상값 찾기: 점수가 100을 넘는(말이 안 되는) 행만 고릅니다.
# 실행하면 예린(550점) 한 행만 보입니다. 4차시에서 배운 조건 검색과 같은 방법입니다.
toy[toy["점수"] > 100]

,이름,점수
3,예린,550.0


### 반도체 데이터 읽고 숫자 요약부터 훑어보기
결측값을 하나씩 세기 전에 `describe()`로 숫자 열 전체를 한 번에 요약합니다(강의 자료 "가장 먼저 하는 습관").
`describe()`는 숫자형 열만 보여줍니다 — `로트번호`·`설비번호`·`공정명` 같은 문자열 열은 나오지 않습니다.
마지막 셀에서 열별 결측 개수를 정확히 셉니다.

In [6]:
# pandas를 불러오고, 오늘 실습 CSV 파일을 읽어 df라는 표(DataFrame)에 저장합니다.
# 실행하면 (113, 9)가 출력됩니다 — 113행 9열짜리 데이터입니다.
import pandas as pd

df = pd.read_csv("../../data/weekly/week05/week05_dirty_process_data.csv")
print(df.shape)  # (중간 확인)

(113, 9)


In [7]:
# describe()로 숫자 열 전체를 한 번에 요약합니다. 결측을 하나씩 세기 전에 가장 먼저 하는 습관입니다.
# 실행하면 숫자 열 5개의 count(값이 있는 개수)·mean(평균)·50%(중앙값)·min·max 등이 표로 나옵니다.
# count가 113보다 작은 열, mean이 50%보다 눈에 띄게 큰 열, min/max가 말이 안 되는 열을 찾아보세요.
print(df.describe())

            온도_섭씨        압력_Pa    가스유량_slm    처리시간_sec        합격여부
count  110.000000   111.000000  110.000000  113.000000  111.000000
mean   304.941298   994.361336   51.350000  119.922124    0.585586
std     33.576979   140.252613   19.933578    5.191128    0.814287
min    291.300000   -42.910520   -5.000000  107.300000   -1.000000
25%    298.100000  1007.850000   48.755000  116.500000    1.000000
50%    300.450000  1013.100000   49.940000  119.800000    1.000000
75%    302.500000  1017.500000   51.377500  123.400000    1.000000
max    557.727366  1033.100000  250.000000  132.300000    1.000000


In [8]:
# 미리 맛보기 1과 같은 방법으로, 반도체 데이터의 열별 결측값 개수를 정확히 셉니다.
# 실행하면 9개 열 이름 옆에 결측 개수가 나옵니다 — describe()의 count가 113보다 작던 열에 결측이 있는지 확인해 보세요.
print(df.isna().sum())

측정시간        0
로트번호        0
설비번호        0
공정명         0
온도_섭씨       3
압력_Pa       2
가스유량_slm    3
처리시간_sec    0
합격여부        2
dtype: int64


## 2단계. 결측값 처리하기
**설명 포인트**: 범주형(합격여부)은 제거, 숫자형은 중앙값 대체라는 원칙을 칠판에 적어둔다.

In [9]:
# 합격여부가 비어 있는 행을 지우고, 그 결과를 다시 df에 저장합니다. 합격여부(1 또는 -1)는 평균·중앙값으로 채울 수 없기 때문입니다.
# 실행하면 (111, 9)가 출력됩니다 — 113행에서 2행이 줄었습니다.
# subset=["열 이름"]: 그 열에 결측이 있는 행만 지우라는 뜻입니다. 앞에 df =를 붙여 다시 저장해야 실제로 반영됩니다.
# 합격여부가 결측인 행은 제거(범주형은 대체 불가)
df = df.dropna(subset=["합격여부"])
print(df.shape)  # (중간 확인)

(111, 9)


In [10]:
# 숫자형 세 열(온도_섭씨, 압력_Pa, 가스유량_slm)의 빈칸을 각 열의 중앙값으로 채웁니다. for 문이 세 열을 하나씩 돌며 같은 일을 반복합니다.
# 실행하면 0이 출력됩니다 — 표 전체에 남은 결측값이 하나도 없다는 뜻입니다.
# .sum().sum(): 첫 번째 sum은 열별 결측 개수, 두 번째 sum은 그 개수들을 모두 더한 총합입니다.
# 숫자형 열은 중앙값으로 대체
for col in ["온도_섭씨", "압력_Pa", "가스유량_slm"]:
    median_value = df[col].median()
    df[col] = df[col].fillna(median_value)

print(df.isna().sum().sum())   # 남은 결측값 총합

0


## 3단계. 중복값 찾고 제거하기

In [11]:
# 모든 열의 값이 완전히 똑같은 행(중복 행)이 몇 개인지 셉니다.
# 실행하면 숫자 하나가 출력됩니다. 중복 여부 함수는 앞에 이미 나온 행과 똑같은 행만 True로 표시하므로, 처음 나온 행은 빼고 "여분"만 세어집니다.
print(df.duplicated().sum())   # 중복 행 개수

3


In [12]:
# 중복 행의 여분을 지우고 다시 df에 저장한 뒤, 표의 크기를 확인합니다.
# 실행하면 (행 개수, 열 개수)가 출력됩니다 — 중복이 3건이면 6행이 아니라 여분 3행만 줄어듭니다(111 → 108).
df = df.drop_duplicates()
print(df.shape)

(108, 9)


## 4단계. 이상값 찾기
**어려워할 부분**: isna()로 이상값이 안 잡힌다는 것을 직접 실행해서 보여준다.

**시연 방법**: 강의 코드대로 고온·음수 압력·비정상 가스유량 세 가지를 모두 찾는다(각 2건). 가스유량 조건의 `|`는 "또는"이다. 이어지는 "직접 해보기" 셀(`df["공정명"].unique()`)로 공정명 오타(`중착`, `포토공정`, `식각 `)를 보여주고, 7단계 "도전 문제 1" 셀로 연결한다. 학생용은 이 셀의 함수 이름이 빈칸이고, 바로 뒤 "참고" 셀이 다시 볼 강의 절을 알려 준다.

In [13]:
# 4차시 조건 검색으로 이상값 세 종류를 각각 찾아 개수를 출력합니다. 이상값은 값이 "있기" 때문에 결측 세기로는 잡히지 않습니다.
# 실행하면 고온 이상값·음수 압력·비정상 가스유량이 각각 몇 건인지 세 줄로 나옵니다.
# |는 "또는"입니다 — 두 조건 중 하나라도 맞으면 고릅니다. 조건마다 괄호 ( )로 감쌉니다.
hot_outlier = df[df["온도_섭씨"] > 400]
neg_pressure = df[df["압력_Pa"] < 0]
bad_gas = df[(df["가스유량_slm"] < 0) | (df["가스유량_slm"] >= 200)]

print("고온 이상값:", len(hot_outlier), "건")
print("음수 압력:", len(neg_pressure), "건")
print("비정상 가스유량:", len(bad_gas), "건")

고온 이상값: 2 건
음수 압력: 2 건
비정상 가스유량: 2 건


### 직접 해보기 — 공정명 열에 어떤 값이 있을까?
강의 자료 "직접 해보기"의 코드입니다. `중착`, `포토공정`처럼 오타가 섞여 있는지 확인해 보세요.

In [14]:
# 공정명 열에 어떤 값들이 들어 있는지, 겹치는 값은 한 번씩만 모아 보여줍니다.
# 실행하면 공정명 목록이 array([...])로 나옵니다 — '중착', '포토공정', 뒤에 공백이 붙은 '식각 '처럼 오타가 보이나요?
df["공정명"].unique()

array(['포토', '산화', '세정', '증착', '식각', '중착', '포토공정', '식각 '], dtype=object)

> **참고**
>
> - 빈칸(함수 이름): 강의 자료 "이상값 찾기 — 4차시 조건 검색 재활용" 절의 "직접 해보기" 상자에 이 코드가 그대로 있습니다. "② 파이썬으로 확인하기 — 숫자 요약부터, 그다음 결측값" 절의 "주의 — describe()는 숫자형 열만 보여준다" 상자에도 같은 함수가 나옵니다.
> - 이 함수는 한 열에 나오는 서로 다른 값을 한 번씩만 모아 보여 줍니다. 문자열 열은 `describe()` 결과에 나오지 않으므로 이렇게 따로 확인합니다.
>
> 정답 해설: `unique`를 채워 `df["공정명"].unique()`로 만든다. 서로 다른 공정명 8개 중 오타 3개(`'중착'`, `'포토공정'`, 뒤에 공백이 붙은 `'식각 '`)가 보이고, 이것을 7단계 "도전 문제 1"에서 고친다.
>
> 예상 결과: `array(['포토', '산화', '세정', '증착', '식각', '중착', '포토공정', '식각 '], dtype=object)`

## 5단계. 이상값 제거하기

**구성**: 행 삭제(강의 "이상값 제거") → `### 참고` loc로 값만 정정(강의 "참고" 절) → `### 체크포인트` reset_index(강의 "체크포인트"). loc 시연은 강의 코드대로 원본을 `sample_df`라는 새 이름으로 다시 읽어서 한다(400도 초과 2건 → 0건, 113행 유지). 정제 중인 `df`는 그대로라 6단계 결과에 영향이 없다. 학생이 `df`로 다시 읽으면 6단계가 `(113행)`으로 나온다.

**설명 포인트**: 여러 값이 동시에 비정상인 행은 행 삭제, 한 열만 튀는 행은 loc 정정이 데이터 손실을 줄인다(정해진 정답은 없음). 실습은 계속 행 삭제 방식으로 이어간다.

In [15]:
# 4단계에서 찾은 이상값의 반대 조건(정상 범위)으로 걸러내, 정상인 행만 다시 df에 저장합니다.
# 실행하면 (행 개수, 열 개수)가 출력됩니다 — 4단계에서 찾은 이상값 행이 빠졌는지 108행과 비교해 보세요.
# &는 "그리고"입니다 — 두 조건이 모두 맞는 행만 남깁니다.
df = df[df["온도_섭씨"] <= 400]
df = df[df["압력_Pa"] >= 0]
df = df[(df["가스유량_slm"] >= 0) & (df["가스유량_slm"] < 200)]

print(df.shape)

(102, 9)


### 참고 — loc로 행을 지우지 않고 "값만" 정정하기
방금은 이상값이 있는 **행 전체**를 지웠습니다. 4차시에서 배운 `loc`를 쓰면 조건에 맞는 행의 **특정 열 값만**
바꿀 수 있습니다(강의 자료 "참고" 절).

이미 400도 초과 행을 지운 `df`에는 바꿀 값이 없으므로, 아래 두 셀은 원본을 `sample_df`라는 **새 이름**으로 다시 읽어
시험합니다. 정제 중인 `df`는 그대로 남으므로 6단계 비교에 영향이 없습니다(같은 이름 `df`로 다시 읽으면 정제 결과가
원본으로 덮여 6단계가 113행 기준으로 계산됩니다). 실습에서는 계속 "행 삭제" 방식을 이어서 씁니다.

In [16]:
# 원본 파일을 sample_df라는 새 이름으로 다시 읽고, 400도를 넘는 칸이 몇 개인지와 행 개수를 확인합니다.
# 정제 중인 df는 건드리지 않으므로 6단계 결과가 바뀌지 않습니다. 실행하면 400도 초과 2건, 행 개수 113이 출력됩니다.
# 위에서 이미 400도 초과 행을 지운 df에는 바꿀 값이 없다.
# 그래서 원본을 sample_df라는 새 이름으로 다시 읽어 시험한다(정제 중인 df는 그대로 둔다).
sample_df = pd.read_csv("../../data/weekly/week05/week05_dirty_process_data.csv")
print("400도 초과:", (sample_df["온도_섭씨"] > 400).sum(), "건 / 행 개수:", len(sample_df))  # (중간 확인)

400도 초과: 2 건 / 행 개수: 113


In [17]:
# sample_df에서 온도가 400도를 넘는 칸만 골라, 그 칸의 값을 온도 중앙값으로 바꿉니다. 행은 지우지 않습니다.
# 실행하면 400도 초과 0건, 행 개수 113이 출력됩니다 — 이상값은 사라지고 행은 그대로 남았습니다.
# df.loc[조건, "열 이름"] = 값: 조건에 맞는 행의 그 열 칸에만 새 값을 씁니다(수정). 다른 칸과 다른 열은 바뀌지 않습니다.
# 온도가 400도를 넘는 행만, 온도_섭씨 값만 중앙값으로 바꾼다(행은 남긴다)
sample_df.loc[sample_df["온도_섭씨"] > 400, "온도_섭씨"] = sample_df["온도_섭씨"].median()
print("400도 초과:", (sample_df["온도_섭씨"] > 400).sum(), "건 / 행 개수:", len(sample_df))

400도 초과: 0 건 / 행 개수: 113


### 체크포인트 — 필터링 후 인덱스에 구멍이 생긴다
`dropna()`·`drop_duplicates()`·조건 검색으로 행을 걸러내면 맨 왼쪽 인덱스 번호가 0, 1, 2, ...로 다시 매겨지지 않고
원래 번호가 그대로 남아 중간중간 번호가 빠집니다. "몇 번째 행"으로 순서대로 접근하고 싶다면
`reset_index(drop=True)`로 번호를 새로 매깁니다(강의 자료 "체크포인트").

In [18]:
# 행을 걸러낸 뒤 df의 마지막 3행을 봅니다.
# 실행하면 맨 왼쪽 인덱스 번호가 101보다 큰 수로 나옵니다 — 행은 102개(0~101)인데, 걸러낸 행의 번호가 빠진 채 원래 번호가 남았기 때문입니다.
# tail(n): 표의 뒤쪽 n행만 보여줍니다(head의 반대).
df.tail(3)  # (중간 확인)

,측정시간,로트번호,설비번호,공정명,온도_섭씨,압력_Pa,가스유량_slm,처리시간_sec,합격여부
107,2024-04-06 16:00,LOT-0331,EQ-04,포토,298.4,1024.5,51.38,128.2,1.0
108,2024-04-07 03:00,LOT-0209,EQ-04,산화,303.1,1005.9,49.45,119.4,1.0
109,03-05-2024 10:00,LOT-0398,EQ-01,세정,301.0,1019.7,51.30,119.0,1.0


In [19]:
# 인덱스 번호를 0부터 새로 매기고 다시 df에 저장한 뒤, 마지막 3행을 다시 봅니다.
# 실행하면 맨 왼쪽 번호가 99, 100, 101로 바뀝니다 — 102행이므로 0~101입니다.
# reset_index(drop=True): 번호를 새로 매기면서, 원래 번호는 새 열로 남기지 않고 버리라는 뜻입니다.
df = df.reset_index(drop=True)
df.tail(3)  # (중간 확인)

,측정시간,로트번호,설비번호,공정명,온도_섭씨,압력_Pa,가스유량_slm,처리시간_sec,합격여부
99,2024-04-06 16:00,LOT-0331,EQ-04,포토,298.4,1024.5,51.38,128.2,1.0
100,2024-04-07 03:00,LOT-0209,EQ-04,산화,303.1,1005.9,49.45,119.4,1.0
101,03-05-2024 10:00,LOT-0398,EQ-01,세정,301.0,1019.7,51.30,119.0,1.0


## 6단계(종합). 정제 전후 평균 비교하기

**예상 결과**: `정제 전 평균 온도: 304.9도 (113행)` / `정제 후 평균 온도: 300.6도 (102행)` — 약 4.3도 차이. 500도가 넘는 고온 이상값 2건이 정제 전 평균을 끌어올렸다는 점을 짚는다.

In [20]:
# 정제 전후의 평균 온도와 행 개수를 나란히 출력해 비교합니다.
# 실행하면 두 줄이 나옵니다 — 정제 후 평균이 정제 전(304.9도)보다 얼마나 내려갔는지 확인해 보세요.
# 정제 전 값 304.9는 1단계 describe()의 온도_섭씨 mean(결측 3칸을 뺀 110개 값의 평균)을 옮겨 적은 것입니다.
# {값:.1f}는 소수점 아래 한 자리까지 보여 달라는 f-string 표기입니다.
print("정제 전 평균 온도: 304.9도 (113행)")   # 정제 전 값(위 describe()의 온도_섭씨 mean을 옮겨 적음)
print(f"정제 후 평균 온도: {df['온도_섭씨'].mean():.1f}도 ({len(df)}행)")

정제 전 평균 온도: 304.9도 (113행)
정제 후 평균 온도: 300.6도 (102행)


## 7단계. 오늘의 학습을 한 문장으로 정리하기
**진행 방법**: 6단계에서 본 정제 전후 평균 숫자를 직접 인용해 말하게 한다. 숫자가 들어간 문장이 훨씬 설득력 있다는 것을 이 자리에서 짚어준다.
**설명 포인트**: 결과 해석 문장 쓰기는 1차시부터 10차시까지 매 차시 빠지지 않는 고정 활동이다(docs/curriculum.md 「차시 간 연결 원칙」). 코드를 친 것으로 끝내지 않고 자기 말로 바꿔 말해보게 하는 것이 이 과정의 목표 — "데이터로 공정 상태를 설명할 수 있는 사람" — 에 직접 닿는 활동이므로 시간이 모자라도 2~3분은 반드시 확보한다.

> (예시 답안) 결측값은 제거하거나 대체하고, 중복값은 drop_duplicates()로, 이상값은 조건 검색으로 찾아 제거한다. 이상값 하나가 평균 같은 요약값을 크게 왜곡할 수 있다.

## 오류 대처 방법
- `KeyError` 발생 시: `df.columns`로 정확한 열 이름 확인.
- fillna 대상 지정 실수: 반드시 `df["열"] = df["열"].fillna(...)` 형태인지 확인.
- 조건 검색 결과가 빈 표로 나올 때: 비교 연산자(`<`, `>`)와 열 이름 철자를 다시 확인.
- 6단계 행 개수가 `(113행)`으로 나올 때: 5단계 "참고 — loc"에서 원본을 `sample_df`가 아니라 `df`로 다시 읽었는지 확인하고, 1단계 파일 읽기 셀부터 다시 실행.
- 도전 문제 2 ②에서 `str[5]`로 쓰면 102행 중 101행이 나온다: 글자 위치는 0부터 센다는 점을 짚는다.

## 확장 실습(빠른 학습자용)
공정명 오타를 str.strip()/replace()로 통일해보게 한다.

아래 "도전 문제 1"(공정명 오타 통일, 두 셀)과 "도전 문제 2"(형식이 다른 측정시간 찾기, 두 셀)는 강의 "도전 실습" 절에 대응하는 **세 노트북 공통 셀**이다. 학생용은 강의 본문에 나온 함수 이름·숫자(`strip`, `replace`, `len`, `4`)를 빈칸으로 가렸고, 빈칸 셀마다 바로 뒤 "참고" 셀이 다시 볼 강의 절을 알려 준다. 도전 문제 2의 ①(글자 수)에서 1건만 나오는 것은 데이터가 그렇게 생긴 것이므로 정정하지 말고 "왜 그럴까?"로 되물어 ②(5번째 글자)로 이어간다(운영안 "빠른 학습자용 추가 과제").

### 도전 문제 1 — 공정명 오타 통일 (빠른 학습자용)
강의 자료 "도전 실습" 절의 도전 문제 1입니다. 4단계 "직접 해보기"에서 본 오타(`중착`, `포토공정`, 뒤에 공백이 붙은 `식각 `)를
두 셀로 나눠 고칩니다 — 먼저 앞뒤 공백을 지우고, 다음에 오타를 올바른 이름으로 바꿉니다. 셀마다 서로 다른 공정명 목록이
어떻게 줄어드는지 확인하세요. 위의 한 문장 정리를 마친 뒤 시간이 남으면 풀어 봅니다.

In [21]:
# 도전 문제 1-①: 공정명 앞뒤에 붙은 공백을 지우고, 그 결과를 다시 df["공정명"]에 저장합니다.
# 실행하면 서로 다른 공정명 목록이 나옵니다 — 4단계 "직접 해보기"에서 본 8개 중 뒤에 공백이 붙어 있던 '식각 '이 '식각'과 합쳐져 7개가 됩니다.
# .str. 뒤에 문자열 함수를 붙이면 열의 모든 칸에 한 번에 적용됩니다.
df["공정명"] = df["공정명"].str.strip()
print(df["공정명"].unique())

['포토' '산화' '세정' '증착' '식각' '중착' '포토공정']


> **참고**
>
> - 빈칸(함수 이름): 강의 자료 "도전 실습" 절의 도전 문제 1 상자에 함수 이름이 두 개 나옵니다. 이 셀에는 그중 **앞의 것**, `str.` 뒤에 붙는 함수를 씁니다. "과제(다음 차시까지)" 절의 과제 2에도 같은 함수가 나옵니다.
> - 이 함수는 글자 **앞뒤**의 공백만 지우고 가운데 글자는 건드리지 않습니다. `'식각 '`처럼 눈에 안 보이는 공백 때문에 서로 다른 값으로 세어지던 것을 하나로 합칠 때 씁니다.
>
> 정답 해설: `strip`을 채워 `df["공정명"].str.strip()`으로 만든다. 모든 칸의 앞뒤 공백이 지워져 `'식각 '`이 `'식각'`과 같은 값이 되므로 서로 다른 공정명이 8개에서 7개로 줄어든다.
>
> 예상 결과: `['포토' '산화' '세정' '증착' '식각' '중착' '포토공정']` (8개 → 7개)

In [22]:
# 도전 문제 1-②: 오타 두 가지를 {옛 값: 새 값} 딕셔너리로 적어 올바른 이름으로 바꾸고, 다시 df["공정명"]에 저장합니다.
# 실행하면 공정명이 포토·산화·세정·증착·식각 5개만 남습니다.
df["공정명"] = df["공정명"].replace({"중착": "증착", "포토공정": "포토"})
print(df["공정명"].unique())

['포토' '산화' '세정' '증착' '식각']


> **참고**
>
> - 빈칸(함수 이름): 강의 자료 "도전 실습" 절의 도전 문제 1 상자에 나오는 두 함수 이름 중 **뒤의 것**입니다. 이 함수는 `str.` 없이 열에 바로 붙입니다.
> - 이 함수에 `{옛 값: 새 값}` 딕셔너리를 넘기면, 옛 값과 **글자까지 완전히 같은** 칸만 새 값으로 바꿉니다. 공백이 붙은 값은 옛 값과 다른 글자로 취급되므로, 1-①에서 공백을 먼저 지운 것입니다.
>
> 정답 해설: `replace`를 채워 `df["공정명"].replace({"중착": "증착", "포토공정": "포토"})`로 만든다. 오타 두 개가 올바른 이름으로 바뀌어 서로 다른 공정명이 5개만 남는다.
>
> 예상 결과: `['포토' '산화' '세정' '증착' '식각']` (7개 → 5개). 이때 `df["공정명"].value_counts()`는 포토 24, 세정 22, 산화 21, 증착 19, 식각 16(합계 102행)이다.

### 도전 문제 2 — 형식이 다른 측정시간 찾기 (빠른 학습자용)
강의 자료 "도전 실습" 절의 도전 문제 2입니다. `측정시간` 열에는 형식이 다른 날짜가 2건(`2024/03/05`, `03-05-2024 10:00`)
섞여 있습니다. ①은 글자 수로, ②는 정해진 위치의 글자로 찾습니다. 두 결과의 건수가 왜 다른지 생각해 보세요.

In [23]:
# 도전 문제 2-①: 측정시간의 글자 수가 16이 아닌 행을 찾습니다. 정상값은 "2024-03-05 10:00"처럼 16자입니다.
# 실행하면 1건(2024/03/05)만 나옵니다 — 코드 실수가 아니라 데이터가 그렇게 생긴 것입니다. 왜 1건뿐인지 다음 셀에서 확인합니다.
df[df["측정시간"].str.len() != 16]

,측정시간,로트번호,설비번호,공정명,온도_섭씨,압력_Pa,가스유량_slm,처리시간_sec,합격여부
98,2024/03/05,LOT-0020,EQ-01,포토,299.7,1005.6,50.48,111.2,1.0


> **참고**
>
> - 빈칸(함수 이름): 강의 자료 "도전 실습" 절의 도전 문제 2 ①에 이 코드가 그대로 있습니다.
> - 이 함수는 문자열 칸마다 글자 수(공백 포함)를 세어 숫자로 돌려줍니다. 그 숫자를 `!= 16`으로 비교하면 16자가 아닌 칸만 True가 되고, 바깥의 `df[ ]`가 4차시 조건 검색처럼 그 행만 골라 줍니다.
>
> 정답 해설: `len`을 채워 `df["측정시간"].str.len()`으로 만든다. 정상값은 16자이고 `2024/03/05`만 10자라 이 1건만 걸린다. `03-05-2024 10:00`은 16자라서 길이로는 걸러지지 않는다.
>
> 예상 결과: 1행(인덱스 98, `2024/03/05`, LOT-0020). 원본 113행 기준 `측정시간` 글자 수 분포는 16자 112건 / 10자 1건이고, 정제 후 102행 기준은 16자 101건 / 10자 1건이다.

In [24]:
# 도전 문제 2-②: 측정시간의 5번째 글자가 "-"가 아닌 행을 찾습니다. 정상값 "2024-03-05 10:00"은 5번째 글자가 "-"입니다.
# 실행하면 2건(2024/03/05, 03-05-2024 10:00)이 모두 나옵니다 — 글자 수가 같아도 형식은 다를 수 있습니다.
# .str[위치 번호]: 칸마다 문자열에서 그 위치의 글자 하나를 꺼냅니다.
df[df["측정시간"].str[4] != "-"]

,측정시간,로트번호,설비번호,공정명,온도_섭씨,압력_Pa,가스유량_slm,처리시간_sec,합격여부
98,2024/03/05,LOT-0020,EQ-01,포토,299.7,1005.6,50.48,111.2,1.0
101,03-05-2024 10:00,LOT-0398,EQ-01,세정,301.0,1019.7,51.30,119.0,1.0


> **참고**
>
> - 빈칸(위치 번호): 강의 자료 "도전 실습" 절의 도전 문제 2 ②에 이 코드와 "글자 위치를 몇부터 세는지" 설명이 함께 있습니다.
> - `.str[위치 번호]`는 칸마다 그 위치의 글자 하나를 꺼냅니다. 파이썬은 위치를 0부터 세므로 첫 번째 글자가 위치 0입니다 — 2차시 리스트 인덱싱과 같은 규칙입니다.
>
> 정답 해설: `4`를 채워 `df["측정시간"].str[4]`로 만든다(5번째 글자 = 위치 4). 정상값은 위치 4가 `-`인데, `2024/03/05`는 `/`, `03-05-2024 10:00`은 `5`라서 두 건이 모두 걸린다.
>
> 예상 결과: 2행(인덱스 98 `2024/03/05`, 인덱스 101 `03-05-2024 10:00`). `str[5]`로 쓰면 102행 중 101행이 나온다.

## 8단계. AI에게 질문하며 더 알아보기
**설명 포인트**: IQR 질문이 나오면, 오늘 배운 "정해진 정상 범위(295~305도)로 찾기"와 "통계적으로
자동 계산하기(IQR)"의 차이를 비교해서 설명해준다.

질문 예시:
- "결측값을 평균으로 채우는 것과 중앙값으로 채우는 것은 언제 다르게 써야 하나요?"
- "이상값(outlier)을 찾는 통계적 방법(IQR, Z-score)에는 어떤 것들이 있나요?"
- "drop_duplicates()에서 특정 열 기준으로만 중복을 판단하려면 어떻게 하나요?"
- "결측값을 무조건 채우거나 지우는 것 말고 다른 처리 방법도 있나요?"

## 9단계. AI에게 코드 생성 요청하고 직접 실행해보기
**설명 포인트**: `quantile()`/IQR은 오늘 배우지 않은 통계 개념이다 — "정상 범위를 사람이 정하는
방식"과 "통계로 자동 계산하는 방식"을 비교해서 설명해준다.

프롬프트 예시: "숫자 데이터에서 IQR(사분위범위) 방법으로 이상값을 찾아주는 코드를 만들어줘."

아래는 AI가 생성해줄 수 있는 예시 코드와 실행 결과다.

In [25]:
# AI가 만들어 줄 수 있는 예시 코드입니다(강의 자료와 같음). IQR(사분위범위)이라는 통계 방법으로 이상값 경계를 자동으로 계산합니다.
# 실행하면 경계를 벗어난 값이 위치 번호와 함께 출력됩니다(5번 위치의 850.0).
# quantile(0.25)/quantile(0.75): 크기순으로 25%·75% 위치의 값입니다(오늘 배우지 않은 함수). 경계 = Q1 - 1.5×IQR ~ Q3 + 1.5×IQR
import pandas as pd

temps = pd.Series([295.4, 298.1, 312.6, 301.0, 296.8, 850.0])

q1, q3 = temps.quantile(0.25), temps.quantile(0.75)
iqr = q3 - q1
lower, upper = q1 - 1.5 * iqr, q3 + 1.5 * iqr

outliers = temps[(temps < lower) | (temps > upper)]
print(outliers)

5    850.0
dtype: float64


## 10단계. 연습문제 — 나만의 정제 리포트 만들기
**설명 포인트**: 1~3단계에서 배운 `isna()`·`duplicated()` 스킬을 하나의 "정제 리포트" 문장으로
요약하는 연습이다. `df`는 이미 여러 단계를 거치며 정제되었으므로 원본과 비교하려면 `raw_df`로
파일을 다시 읽어야 한다는 점을 짚어준다 — 실무에서도 이런 "정제 보고서"를 남기는 습관이 중요하다.

In [26]:
# 원본 파일을 raw_df라는 새 이름으로 다시 읽어, 합격여부 결측 개수와 전체 중복 행 개수를 세고 한 문장 리포트로 출력합니다.
# df는 이미 정제가 끝났으므로 원본과 비교하려면 파일을 새로 읽어야 합니다. 실행하면 "정제 리포트: 결측 ○건, 중복 ○건을 확인했습니다." 한 줄이 나옵니다.
raw_df = pd.read_csv("../../data/weekly/week05/week05_dirty_process_data.csv")
missing_count = raw_df["합격여부"].isna().sum()
duplicate_count = raw_df.duplicated().sum()

print(f"정제 리포트: 결측 {missing_count}건, 중복 {duplicate_count}건을 확인했습니다.")

정제 리포트: 결측 2건, 중복 3건을 확인했습니다.


In [27]:
# 지금까지 정제한 df의 최종 행 개수를 리포트에 한 줄 덧붙입니다(강의 정답 박스에는 없는, 노트북에만 있는 줄).
# 실행하면 "정제 후 최종 데이터는 ○행입니다."가 나옵니다 — 5단계 결과와 같은지 확인해 보세요.
print(f"정제 후 최종 데이터는 {len(df)}행입니다.")

정제 후 최종 데이터는 102행입니다.


## 11단계. AI로 재미있는 미니 프로그램 만들기 🎉 — 이상값 탐정 게임
**설명 포인트**: 오늘 배운 "정상 범위를 벗어난 값 찾기"를 통계(평균 ± 표준편차)로 재미있게
재구성한 것이다. `statistics` 모듈은 아직 배우지 않았지만, 온도_섭씨 이상값을 찾던 방식(조건
검색)과 원리가 같다는 점을 연결해서 설명하면 좋다.

프롬프트 예시: "숫자 리스트를 받아서, 평균에서 표준편차의 2배 이상 벗어난 값을 '용의자'로 찾아
🔍 이모지와 함께 출력하는 파이썬 코드를 만들어줘."

아래는 AI가 생성해줄 수 있는 예시 코드다.

**예상 결과**: 예시 값 8개가 모두 `→ 정상 범위`로 나오고 용의자는 하나도 잡히지 않는다. 극단값(612.3, 15.0)이 표준편차 자체를 크게 늘려 2배 기준을 벗어나지 못하기 때문이다(강의 "더 놀아보기"). 학생이 "코드가 틀렸다"고 하면 이 점을 AI에게 되물어 기준(2배→3배)이나 방법(IQR)을 바꿔 보게 한다.

In [28]:
# AI가 만들어 줄 수 있는 "이상값 탐정 게임" 예시 코드입니다(강의 자료와 같음). 평균에서 표준편차의 2배보다 멀리 떨어진 값을 "용의자"로 찾습니다.
# 실행하면 8개 값이 한 줄씩 판정됩니다 — 이 예시 값에서는 612.3, 15.0 같은 극단값이 표준편차 자체를 키워 용의자가 하나도 잡히지 않습니다.
# statistics.mean/stdev: 평균과 표준편차(값이 흩어진 정도)를 계산합니다. abs(): 음수를 양수로 바꾼 절댓값입니다.
import statistics

sample_values = [301.2, 298.5, 305.0, 612.3, 299.9, 300.1, 15.0, 302.4]

mean_v = statistics.mean(sample_values)
std_v = statistics.stdev(sample_values)

print("🕵️ 이상값 탐정 게임 시작!")
for v in sample_values:
    if abs(v - mean_v) > 2 * std_v:
        print(f"🔍 용의자 발견! {v} (평균에서 많이 벗어났어요)")
    else:
        print(f"{v} → 정상 범위")

🕵️ 이상값 탐정 게임 시작!
301.2 → 정상 범위
298.5 → 정상 범위
305.0 → 정상 범위
612.3 → 정상 범위
299.9 → 정상 범위
300.1 → 정상 범위
15.0 → 정상 범위
302.4 → 정상 범위
